src/main/java/org/jabref/gui/WaitForBackgroundtasksFinishedDialog.java

```diff
@@ -0,0 +1,49 @@
+package org.jabref.gui;
+
+import javafx.concurrent.Task;
+import org.jabref.logic.l10n.Localization;
+
+import java.util.List;
+
+/**
+ * Dialog shown when closing of application needs to wait for some background tasks.
+ */
+public class WaitForBackgroundtasksFinishedDialog {
+
+    private final DialogService dialogService;
+
+    public WaitForBackgroundtasksFinishedDialog(DialogService dialogService) {
+        this.dialogService = dialogService;
+    }
+
+    public void showAndWait(StateManager stateManager) {
+        if (stateManager.anyTaskRunningBinding.getValue()) {
+            Task<Void> waitForBackgroundtasksFinished = new Task<Void>() {
+                @Override
+                protected Void call() throws Exception {
+                    System.out.println("THREAD STARTED");
+                    while (stateManager.anyTaskRunningBinding.getValue()) {
+                        System.out.println("updated value to " + stateManager.tasksProgressBinding.getValue());
+                        updateProgress(stateManager.tasksProgressBinding.getValue(), 1);
+                        if (isCancelled()) {
+                            return null;
+                        } else {
+                            Thread.sleep(100);
+                        }
+                    }
+                    return null;
+                }
+            };
+
+            Thread th = new Thread(waitForBackgroundtasksFinished);
+            th.setDaemon(true);
+            th.start();
+
+            dialogService.showProgressDialogAndWait(
+                    Localization.lang("Please wait..."),
+                    Localization.lang("Waiting for background tasks to finish") + "...",
+                    waitForBackgroundtasksFinished
+            );
```

Comment:
Add a log statement here as well to make sure that the progress dialog does indeed wait?

Reply:
I currently have a log message after the loop to see whether the isCancelled property or my tasks-running property caused the stop. I now also put a log message where you suggested, just after the wait. They are exectued in the correct order. So yes, the wait does work, but the task exits too early.

Reply:
I now also get a warning about the binding:

> May 10, 2020 12:02:50 AM javafx.beans.binding.Bindings$1 computeValue
WARNING: Exception while evaluating binding
java.lang.IllegalStateException: Task must only be used from the FX Application Thread
        at javafx.graphics/javafx.concurrent.Task.checkThread(Task.java:1220)
        at javafx.graphics/javafx.concurrent.Task.isRunning(Task.java:987)
        at java.base/java.util.stream.MatchOps$1MatchSink.accept(MatchOps.java:90)
        at java.base/java.util.AbstractList$RandomAccessSpliterator.tryAdvance(AbstractList.java:706)
        at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
        at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
        at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
        at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:230)
        at java.base/java.util.stream.MatchOps$MatchOp.evaluateSequential(MatchOps.java:196)
        at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.base/java.util.stream.ReferencePipeline.anyMatch(ReferencePipeline.java:528)
        at org.jabref/org.jabref.gui.StateManager.lambda$new$1(StateManager.java:132)
        at javafx.base/javafx.beans.binding.Bindings$1.computeValue(Bindings.java:157)
        at javafx.base/javafx.beans.binding.BooleanBinding.get(BooleanBinding.java:155)
        at org.jabref/org.jabref.gui.WaitForBackgroundtasksFinishedDialog$1.call(WaitForBackgroundtasksFinishedDialog.java:28)
        at org.jabref/org.jabref.gui.WaitForBackgroundtasksFinishedDialog$1.call(WaitForBackgroundtasksFinishedDialog.java:24)
        at javafx.graphics/javafx.concurrent.Task$TaskCallable.call(Task.java:1425)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.lang.Thread.run(Thread.java:832)

I tried to run the task on the Javafx thread by using:

`DefaultTaskExecutor.runInJavaFXThread(waitForBackgroundtasksFinished);`

The loop is then run more often (I guess until the download is finished), but the gui is frozen because of the task.

Reply:
The problem was that the extractor I added to the task list in StateManager did not include the isRunning property. It still did not run perfectly and was not pretty, therefore I moved to a custom dialog which works fine.